In [1]:
from langchain_google_genai import GoogleGenerativeAIEmbeddings
from dotenv import load_dotenv
from sklearn.metrics.pairwise import cosine_similarity
import numpy as np


d:\Gen AI\venv\lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [2]:
load_dotenv()

embedding = GoogleGenerativeAIEmbeddings(model='gemini-embedding-001', dimensions=300)

documents = [
    "Virat Kohli is an Indian cricketer known for his aggressive batting and leadership.",
    "MS Dhoni is a former Indian captain famous for his calm demeanor and finishing skills.",
    "Sachin Tendulkar, also known as the 'God of Cricket', holds many batting records.",
    "Rohit Sharma is known for his elegant batting and record-breaking double centuries.",
    "Jasprit Bumrah is an Indian fast bowler known for his unorthodox action and yorkers."
]


In [3]:
query = 'tell me about bumrah'

doc_embeddings = embedding.embed_documents(documents)
query_embedding = embedding.embed_query(query)

In [4]:
cosine_similarity([query_embedding], doc_embeddings) #2Darray of cosine similarity scores between the query and each document

array([[0.6382609 , 0.61644691, 0.62209685, 0.63079128, 0.74880076]])

In [5]:
scores = cosine_similarity([query_embedding], doc_embeddings)[0]
scores

array([0.6382609 , 0.61644691, 0.62209685, 0.63079128, 0.74880076])

- **Explanation of using enumarate and lambda function**
scores = [0.616, 0.622, 0.631, 0.638, 0.749]

enumerate(scores) produces:
(0, 0.616)  ← tuple: (index, value)
(1, 0.622)
(2, 0.631)
(3, 0.638)
(4, 0.749)

lambda x : x[1]
   ↑    ↑    ↑
  name param extract 2nd element

x = (0, 0.616)
x[0] = 0      ← first element (index)
x[1] = 0.616  ← second element (value) ← THIS IS USED FOR SORTING!

**Note** - When you sort by x[1], Python:

Extracts the value at index 1 from each tuple
Compares only those values
Keeps the entire tuple together when reordering

In [7]:
sorted(list(enumerate(scores)),key=lambda x:x[1])

[(1, np.float64(0.6164469114007525)),
 (2, np.float64(0.6220968497226282)),
 (3, np.float64(0.630791277002617)),
 (0, np.float64(0.6382608988384979)),
 (4, np.float64(0.7488007610138675))]

In [8]:
index, score = sorted(list(enumerate(scores)),key=lambda x:x[1])[-1]
print(index, score)

4 0.7488007610138675


In [9]:
query = 'tell me about bumrah'

doc_embeddings = embedding.embed_documents(documents)
query_embedding = embedding.embed_query(query)

scores = cosine_similarity([query_embedding], doc_embeddings)[0]

index, score = sorted(list(enumerate(scores)),key=lambda x:x[1])[-1]

print(query)
print(documents[index])
print("similarity score is:", score)

tell me about bumrah
Jasprit Bumrah is an Indian fast bowler known for his unorthodox action and yorkers.
similarity score is: 0.7488007610138675
